File: internal/session/list.go
package session

import (
	"context"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"unicode"

	"github.com/baiyuqing/otto/internal/model"
	"golang.org/x/sys/unix"
)

const (
	maxListSessions        = 20
	maxSessionPreviewRunes = 120
	maxSkippedSessionCount = int(^uint(0) >> 1)
)

type listCandidate struct {
	name string
	path string
	info os.FileInfo
}

func Inspect(ctx context.Context, path string) (SessionInfo, []Warning, error) {
	if err := ctx.Err(); err != nil {
		return SessionInfo{}, nil, err
	}

	file, info, err := openSessionFileReadOnlyNoFollow(path)
	if err != nil {
		return SessionInfo{}, nil, err
	}
	defer file.Close()

	return inspectOpenedSession(ctx, path, file, info)
}

func List(ctx context.Context, root, workspace, currentPath string, limit int) (ListResult, error) {
	if limit < 1 || limit > maxListSessions {
		return ListResult{}, fmt.Errorf("list limit must be between 1 and %d", maxListSessions)
	}
	if err := ctx.Err(); err != nil {
		return ListResult{}, err
	}

	workspaceCanonical, err := canonicalWorkspace(workspace)
	if err != nil {
		return ListResult{}, err
	}

	rootDir, err := openSessionRootNoFollow(root)
	if err != nil {
		return ListResult{}, err
	}
	defer rootDir.Close()

	sessionDir, directory, exists, err := openWorkspaceSessionDirectoryNoFollow(rootDir, root, workspaceCanonical)
	if err != nil {
		return ListResult{}, err
	}
	if !exists {
		return ListResult{}, nil
	}
	defer sessionDir.Close()

	entries, err := sessionDir.ReadDir(-1)
	if err != nil {
		return ListResult{}, fmt.Errorf("read session directory: %w", err)
	}

	currentCanonical := ""
	var currentInfo os.FileInfo
	haveCurrentInfo := false
	if currentPath != "" {
		currentCanonical, err = canonicalWorkspace(currentPath)
		if err != nil {
			return ListResult{}, err
		}
		currentInfo, err = os.Stat(currentPath)
		if err == nil {
			haveCurrentInfo = true
		}
	}

	result := ListResult{}
	candidates := make([]listCandidate, 0, len(entries))
	for _, entry := range entries {
		if err := ctx.Err(); err != nil {
			return ListResult{}, err
		}
		if filepath.Ext(entry.Name()) != ".jsonl" {
			continue
		}
		file, info, path, err := openSessionFileReadOnlyNoFollowAt(sessionDir, directory, entry.Name())
		if err != nil {
			result.Skipped = incrementSkipped(result.Skipped)
			continue
		}
		if closeErr := file.Close(); closeErr != nil {
			result.Skipped = incrementSkipped(result.Skipped)
			continue
		}
		candidates = append(candidates, listCandidate{name: entry.Name(), path: path, info: info})
	}

	sort.Slice(candidates, func(i, j int) bool {
		left, right := candidates[i], candidates[j]
		if left.info.ModTime().Equal(right.info.ModTime()) {
			return left.path > right.path
		}
		return left.info.ModTime().After(right.info.ModTime())
	})

	sessions := make([]SessionInfo, 0, min(limit, len(candidates)))
	for _, candidate := range candidates {
		if len(sessions) == limit {
			break
		}
		if err := ctx.Err(); err != nil {
			return ListResult{}, err
		}
		file, info, path, err := openSessionFileReadOnlyNoFollowAt(sessionDir, directory, candidate.name)
		if err != nil {
			result.Skipped = incrementSkipped(result.Skipped)
			continue
		}
		if !sameListCandidateMetadata(candidate.info, info) {
			_ = file.Close()
			result.Skipped = incrementSkipped(result.Skipped)
			continue
		}

		sessionInfo, _, inspectErr := inspectOpenedSession(ctx, path, file, info)
		closeErr := file.Close()
		if inspectErr != nil || closeErr != nil {
			result.Skipped = incrementSkipped(result.Skipped)
			continue
		}

		candidateWorkspace, err := canonicalWorkspace(sessionInfo.CWD)
		if err != nil || candidateWorkspace != workspaceCanonical {
			result.Skipped = incrementSkipped(result.Skipped)
			continue
		}

		if currentCanonical != "" {
			if haveCurrentInfo {
				sessionInfo.Current = os.SameFile(info, currentInfo)
			} else {
				candidateCanonical, err := canonicalWorkspace(path)
				if err != nil {
					result.Skipped = incrementSkipped(result.Skipped)
					continue
				}
				sessionInfo.Current = candidateCanonical == currentCanonical
			}
		}
		sessions = append(sessions, sessionInfo)
	}

	result.Sessions = sessions
	return result, nil
}

func sessionDirectory(root, workspace string) (string, error) {
	key, err := workspaceKey(workspace)
	if err != nil {
		return "", err
	}
	return filepath.Join(root, key), nil
}

func openSessionRootNoFollow(root string) (*os.File, error) {
	file, err := openPathNoFollow(root, unix.O_RDONLY|unix.O_DIRECTORY, 0)
	if err != nil {
		switch {
		case err == unix.ELOOP:
			return nil, fmt.Errorf("%w: session root is a symlink", ErrInvalidSession)
		case err == unix.ENOTDIR:
			return nil, fmt.Errorf("%w: session root is not a directory", ErrInvalidSession)
		default:
			return nil, fmt.Errorf("open session root: %w", err)
		}
	}
	return file, nil
}

func openWorkspaceSessionDirectoryNoFollow(rootDir *os.File, root, workspace string) (*os.File, string, bool, error) {
	key, err := workspaceKey(workspace)
	if err != nil {
		return nil, "", false, err
	}
	path := filepath.Join(root, key)
	file, err := openPathAtNoFollow(rootDir, key, unix.O_RDONLY|unix.O_DIRECTORY, 0)
	if err != nil {
		switch {
		case err == unix.ENOENT:
			return nil, path, false, nil
		case err == unix.ELOOP:
			return nil, "", false, fmt.Errorf("%w: session directory is a symlink", ErrInvalidSession)
		case err == unix.ENOTDIR:
			return nil, "", false, fmt.Errorf("%w: session directory is not a directory", ErrInvalidSession)
		default:
			return nil, "", false, fmt.Errorf("open session directory: %w", err)
		}
	}
	return file, path, true, nil
}

func openSessionFileReadOnlyNoFollow(path string) (*os.File, os.FileInfo, error) {
	file, err := openPathNoFollow(path, unix.O_RDONLY, 0)
	if err != nil {
		if err == unix.ELOOP {
			return nil, nil, fmt.Errorf("%w: session file is a symlink", ErrInvalidSession)
		}
		return nil, nil, fmt.Errorf("open session file: %w", err)
	}
	info, err := file.Stat()
	if err != nil {
		_ = file.Close()
		return nil, nil, fmt.Errorf("stat session file: %w", err)
	}
	if !info.Mode().IsRegular() {
		_ = file.Close()
		return nil, nil, fmt.Errorf("%w: session file is not a regular file", ErrInvalidSession)
	}
	return file, info, nil
}

func openSessionFileReadOnlyNoFollowAt(directory *os.File, directoryPath, name string) (*os.File, os.FileInfo, string, error) {
	path := filepath.Join(directoryPath, name)
	file, err := openPathAtNoFollow(directory, name, unix.O_RDONLY, 0)
	if err != nil {
		if err == unix.ELOOP {
			return nil, nil, "", fmt.Errorf("%w: session file is a symlink", ErrInvalidSession)
		}
		return nil, nil, "", fmt.Errorf("open session file: %w", err)
	}
	info, err := file.Stat()
	if err != nil {
		_ = file.Close()
		return nil, nil, "", fmt.Errorf("stat session file: %w", err)
	}
	if !info.Mode().IsRegular() {
		_ = file.Close()
		return nil, nil, "", fmt.Errorf("%w: session file is not a regular file", ErrInvalidSession)
	}
	return file, info, path, nil
}

func openPathNoFollow(path string, flags int, perm uint32) (*os.File, error) {
	fd, err := unix.Open(path, flags|unix.O_CLOEXEC|unix.O_NOFOLLOW, perm)
	if err != nil {
		return nil, err
	}
	return fileFromFD(fd, path)
}

func openPathAtNoFollow(directory *os.File, name string, flags int, perm uint32) (*os.File, error) {
	fd, err := unix.Openat(int(directory.Fd()), name, flags|unix.O_CLOEXEC|unix.O_NOFOLLOW, perm)
	if err != nil {
		return nil, err
	}
	return fileFromFD(fd, filepath.Join(directory.Name(), name))
}

func fileFromFD(fd int, name string) (*os.File, error) {
	file := os.NewFile(uintptr(fd), name)
	if file == nil {
		_ = unix.Close(fd)
		return nil, fmt.Errorf("create file handle for %s: invalid descriptor", name)
	}
	return file, nil
}

func inspectOpenedSession(ctx context.Context, path string, file *os.File, info os.FileInfo) (SessionInfo, []Warning, error) {
	if err := rejectOversizedSessionFile(file); err != nil {
		return SessionInfo{}, nil, err
	}

	decoded, err := decodePiFileReadOnlyContext(ctx, file)
	if err != nil {
		return SessionInfo{}, nil, err
	}
	if err := ctx.Err(); err != nil {
		return SessionInfo{}, nil, err
	}

	createdAt, err := validatePiHeader(decoded.Header)
	if err != nil {
		return SessionInfo{}, nil, err
	}
	leafID := ""
	if len(decoded.Entries) > 0 {
		leafID = decoded.Entries[len(decoded.Entries)-1].ID
	}
	resolved, warnings, err := buildContext(decoded.Entries, leafID)
	if err != nil {
		return SessionInfo{}, nil, err
	}
	preview := previewText(lastUserText(resolved.Messages))
	name := previewText(resolved.SessionName)
	if name == "" {
		name = preview
	}

	return SessionInfo{
		Path:         path,
		ID:           decoded.Header.ID,
		CWD:          decoded.Header.CWD,
		Name:         name,
		Created:      createdAt,
		Modified:     info.ModTime(),
		MessageCount: len(resolved.Messages),
		LastUserText: preview,
		Profile:      resolved.Runtime.Profile,
		Provider:     resolved.Runtime.Provider,
		Model:        resolved.Runtime.Model,
	}, warnings, nil
}

func decodePiFileReadOnlyContext(ctx context.Context, file *os.File) (piFile, error) {
	if err := ctx.Err(); err != nil {
		return piFile{}, err
	}
	finalStart, _, incomplete, err := finalPiRecordState(file)
	if err != nil {
		return piFile{}, err
	}
	if err := ctx.Err(); err != nil {
		return piFile{}, err
	}
	if incomplete && finalStart > 0 {
		decoded, _, err := decodePiFile(&contextReader{ctx: ctx, reader: io.NewSectionReader(file, 0, finalStart)})
		return decoded, err
	}
	if _, err := file.Seek(0, io.SeekStart); err != nil {
		return piFile{}, fmt.Errorf("seek session file: %w", err)
	}
	decoded, _, err := decodePiFile(&contextReader{ctx: ctx, reader: file})
	return decoded, err
}

type contextReader struct {
	ctx    context.Context
	reader io.Reader
}

func (r *contextReader) Read(buffer []byte) (int, error) {
	if err := r.ctx.Err(); err != nil {
		return 0, err
	}
	count, err := r.reader.Read(buffer)
	if err == nil && r.ctx.Err() != nil {
		return count, r.ctx.Err()
	}
	return count, err
}

func lastUserText(messages []model.Message) string {
	for index := len(messages) - 1; index >= 0; index-- {
		if messages[index].Role == model.RoleUser {
			return messages[index].Text()
		}
	}
	return ""
}

func previewText(text string) string {
	var builder strings.Builder
	builder.Grow(maxSessionPreviewRunes*6 + 3)
	used := 0
	wroteContent := false
	pendingSpace := false
	truncated := false

	for _, r := range text {
		if unicode.IsSpace(r) {
			pendingSpace = wroteContent
			continue
		}

		needed := previewRuneWidth(r)
		if pendingSpace {
			needed++
		}
		if used+needed > maxSessionPreviewRunes {
			truncated = true
			break
		}
		if pendingSpace {
			builder.WriteByte(' ')
			used++
			pendingSpace = false
		}
		writePreviewRune(&builder, r)
		used += previewRuneWidth(r)
		wroteContent = true
	}

	if !wroteContent {
		return ""
	}
	if !truncated {
		return builder.String()
	}
	return builder.String() + "..."
}

func previewRuneWidth(r rune) int {
	switch {
	case r < 0x100 && unicode.IsControl(r):
		return 4
	case unicode.IsControl(r):
		return 2 + max(4, hexWidth(uint32(r)))
	default:
		return 1
	}
}

func writePreviewRune(builder *strings.Builder, r rune) {
	switch {
	case r < 0x100 && unicode.IsControl(r):
		builder.WriteString(`\x`)
		writeHex(builder, uint32(r), 2)
	case unicode.IsControl(r):
		builder.WriteString(`\u`)
		writeHex(builder, uint32(r), max(4, hexWidth(uint32(r))))
	default:
		builder.WriteRune(r)
	}
}

func hexWidth(value uint32) int {
	width := 1
	for value >= 16 {
		value /= 16
		width++
	}
	return width
}

func writeHex(builder *strings.Builder, value uint32, width int) {
	const digits = "0123456789abcdef"
	var buffer [8]byte
	for index := width - 1; index >= 0; index-- {
		buffer[index] = digits[value&0xf]
		value >>= 4
	}
	for _, digit := range buffer[:width] {
		builder.WriteByte(digit)
	}
}

func sameListCandidateMetadata(expected, current os.FileInfo) bool {
	return expected != nil && current != nil &&
		os.SameFile(expected, current) &&
		expected.Mode() == current.Mode() &&
		expected.Size() == current.Size() &&
		expected.ModTime().Equal(current.ModTime())
}

func incrementSkipped(count int) int {
	if count < maxSkippedSessionCount {
		return count + 1
	}
	return count
}

File: crates/kite/src/session/list.rs
//! Listing and inspecting session files.
//!
//! Ownership: every function opens and closes its own descriptors. Concurrency:
//! stateless, safe to call from any thread. Errors: a candidate that cannot be
//! read is counted in [`ListResult::skipped`] and never surfaced; only a
//! failure of the listing itself is returned.

use std::fs::{File, Metadata};
use std::path::{Path, PathBuf};

use chrono::{DateTime, Utc};
use kite_core::session::context::{last_user_text, preview_text};
use kite_core::session::{ListResult, PiError, SessionInfo, Warning, build_context};

use super::fsops::{self, Dir};
use super::store::{decode_pi_file_read_only, reject_oversized_session_file, validate_pi_header};

/// The most sessions one [`list`] call may return.
pub const MAX_LIST_SESSIONS: usize = 20;

/// The directory a workspace's sessions live in: the root plus the workspace
/// key. The directory need not exist.
pub fn session_directory(root: &Path, workspace: &str) -> Result<PathBuf, PiError> {
    Ok(root.join(fsops::workspace_key(Path::new(workspace))?))
}

/// Reads one session file's listing row without opening a store.
pub fn inspect(path: &Path) -> Result<(SessionInfo, Vec<Warning>), PiError> {
    let (mut file, metadata) = open_session_file_read_only_no_follow(path)?;
    inspect_opened_session(&path.to_string_lossy(), &mut file, &metadata)
}

/// The workspace's sessions, newest first, capped at `limit`.
///
/// `current_path` marks the row for the session already open, by file
/// identity when it can be stat'ed and by canonical path otherwise. A
/// candidate that cannot be opened, changed between the two opens, belongs to
/// a different workspace, or shares a header id with a row already listed
/// (kept from the newest-modified file) is counted as skipped.
pub fn list(
    root: &Path,
    workspace: &str,
    current_path: &str,
    limit: usize,
) -> Result<ListResult, PiError> {
    if !(1..=MAX_LIST_SESSIONS).contains(&limit) {
        return Err(PiError::other(format!(
            "list limit must be between 1 and {MAX_LIST_SESSIONS}"
        )));
    }
    let workspace_canonical = fsops::canonical_workspace(Path::new(workspace))?;
    let root_dir = open_session_root_no_follow(root)?;
    let Some((session_dir, directory)) =
        open_workspace_session_directory_no_follow(&root_dir, root, &workspace_canonical)?
    else {
        return Ok(ListResult::default());
    };

    let mut result = ListResult::default();
    let mut candidates: Vec<(String, String, Metadata)> = Vec::new();
    for name in read_directory_names(&directory)? {
        if !name.ends_with(".jsonl") {
            continue;
        }
        match open_session_file_read_only_no_follow_at(&session_dir, &directory, &name) {
            Ok((_, metadata, path)) => candidates.push((name, path, metadata)),
            Err(_) => result.skipped = increment_skipped(result.skipped),
        }
    }

    candidates.sort_by(|left, right| {
        let (left_time, right_time) = (modified_time(&left.2), modified_time(&right.2));
        right_time
            .cmp(&left_time)
            .then_with(|| right.1.cmp(&left.1))
    });

    let current_canonical = (!current_path.is_empty())
        .then(|| fsops::canonical_workspace(Path::new(current_path)))
        .transpose()?;
    let current_metadata = std::fs::metadata(current_path).ok();

    let mut sessions = Vec::new();
    let mut seen_ids = std::collections::HashSet::new();
    for (name, _, expected) in &candidates {
        if sessions.len() == limit {
            break;
        }
        let Ok((mut file, metadata, path)) =
            open_session_file_read_only_no_follow_at(&session_dir, &directory, name)
        else {
            result.skipped = increment_skipped(result.skipped);
            continue;
        };
        if !same_list_candidate_metadata(expected, &metadata) {
            result.skipped = increment_skipped(result.skipped);
            continue;
        }
        let Ok((mut info, _)) = inspect_opened_session(&path, &mut file, &metadata) else {
            result.skipped = increment_skipped(result.skipped);
            continue;
        };
        match fsops::canonical_workspace(Path::new(&info.cwd)) {
            Ok(candidate) if candidate == workspace_canonical => {}
            _ => {
                result.skipped = increment_skipped(result.skipped);
                continue;
            }
        }
        if !seen_ids.insert(info.id.clone()) {
            result.skipped = increment_skipped(result.skipped);
            continue;
        }
        if let Some(current_canonical) = current_canonical.as_deref() {
            info.current = match current_metadata.as_ref() {
                Some(current) => fsops::file_identity(&metadata) == fsops::file_identity(current),
                None => match fsops::canonical_workspace(Path::new(&path)) {
                    Ok(candidate) => candidate == current_canonical,
                    Err(_) => {
                        result.skipped = increment_skipped(result.skipped);
                        continue;
                    }
                },
            };
        }
        sessions.push(info);
    }
    result.sessions = sessions;
    Ok(result)
}

/// The names in `directory`, unsorted.
///
/// The directory is read by path rather than through the no-follow descriptor.
/// Names are only a work list: every file is still opened with `openat` on
/// that descriptor, so a directory swapped between the two steps yields opens
/// that fail and rows that are skipped.
fn read_directory_names(directory: &str) -> Result<Vec<String>, PiError> {
    let entries = std::fs::read_dir(directory)
        .map_err(|error| PiError::other(format!("read session directory: {error}")))?;
    let mut names = Vec::new();
    for entry in entries {
        let entry =
            entry.map_err(|error| PiError::other(format!("read session directory: {error}")))?;
        names.push(entry.file_name().to_string_lossy().into_owned());
    }
    Ok(names)
}

/// Opens the session root, refusing a symlink or a non-directory.
pub(crate) fn open_session_root_no_follow(root: &Path) -> Result<Dir, PiError> {
    fsops::open_dir_no_follow(root).map_err(|error| {
        if fsops::is_eloop(&error) {
            PiError::invalid("session root is a symlink")
        } else if fsops::is_enotdir(&error) {
            PiError::invalid("session root is not a directory")
        } else {
            PiError::other(format!("open session root: {error}"))
        }
    })
}

/// Opens the workspace's session directory under `root_dir`. `Ok(None)` means
/// the workspace has no sessions yet.
pub(crate) fn open_workspace_session_directory_no_follow(
    root_dir: &Dir,
    root: &Path,
    workspace_canonical: &str,
) -> Result<Option<(Dir, String)>, PiError> {
    let key = fsops::workspace_key_of_canonical(workspace_canonical);
    let path = root.join(&key).to_string_lossy().into_owned();
    match fsops::open_dir_at_no_follow(root_dir, &key) {
        Ok(dir) => Ok(Some((dir, path))),
        Err(error) if fsops::is_enoent(&error) => Ok(None),
        Err(error) if fsops::is_eloop(&error) => {
            Err(PiError::invalid("session directory is a symlink"))
        }
        Err(error) if fsops::is_enotdir(&error) => {
            Err(PiError::invalid("session directory is not a directory"))
        }
        Err(error) => Err(PiError::other(format!("open session directory: {error}"))),
    }
}

/// Opens one session file read-only, refusing a symlink or a non-regular file.
pub(crate) fn open_session_file_read_only_no_follow(
    path: &Path,
) -> Result<(File, Metadata), PiError> {
    let file = fsops::open_no_follow(path, libc::O_RDONLY).map_err(|error| {
        if fsops::is_eloop(&error) {
            PiError::invalid("session file is a symlink")
        } else {
            PiError::other(format!("open session file: {error}"))
        }
    })?;
    let metadata = regular_file_metadata(&file)?;
    Ok((file, metadata))
}

pub(crate) fn open_session_file_read_only_no_follow_at(
    directory: &Dir,
    directory_path: &str,
    name: &str,
) -> Result<(File, Metadata, String), PiError> {
    let path = Path::new(directory_path)
        .join(name)
        .to_string_lossy()
        .into_owned();
    let file = fsops::open_at_no_follow(directory, name, libc::O_RDONLY).map_err(|error| {
        if fsops::is_eloop(&error) {
            PiError::invalid("session file is a symlink")
        } else {
            PiError::other(format!("open session file: {error}"))
        }
    })?;
    let metadata = regular_file_metadata(&file)?;
    Ok((file, metadata, path))
}

fn regular_file_metadata(file: &File) -> Result<Metadata, PiError> {
    let metadata = file
        .metadata()
        .map_err(|error| PiError::other(format!("stat session file: {error}")))?;
    if !metadata.is_file() {
        return Err(PiError::invalid("session file is not a regular file"));
    }
    Ok(metadata)
}

/// Builds the listing row for an already-opened session file.
pub(crate) fn inspect_opened_session(
    path: &str,
    file: &mut File,
    metadata: &Metadata,
) -> Result<(SessionInfo, Vec<Warning>), PiError> {
    reject_oversized_session_file(file)?;
    let decoded = decode_pi_file_read_only(file)?;
    let created_at = validate_pi_header(&decoded.header)?;
    let leaf_id = decoded
        .entries
        .last()
        .map(|entry| entry.id.clone())
        .unwrap_or_default();
    let (resolved, warnings) = build_context(&decoded.entries, &leaf_id)?;
    let preview = preview_text(&last_user_text(&resolved.messages));
    let name = match preview_text(&resolved.session_name) {
        name if name.is_empty() => preview.clone(),
        name => name,
    };
    Ok((
        SessionInfo {
            path: path.to_owned(),
            id: decoded.header.id.clone(),
            cwd: decoded.header.cwd.clone(),
            name,
            created: created_at,
            modified: modified_time(metadata),
            message_count: resolved.messages.len() as i64,
            last_user_text: preview,
            profile: resolved.runtime.profile,
            provider: resolved.runtime.provider,
            model: resolved.runtime.model,
            thinking: pi_level_to_thinking(&resolved.thinking_level)?,
            current: false,
        },
        warnings,
    ))
}

fn pi_level_to_thinking(thinking: &str) -> Result<String, PiError> {
    match thinking {
        "" | "off" => Ok(String::new()),
        "low" | "medium" | "high" | "xhigh" | "max" => Ok(thinking.to_string()),
        _ => Err(PiError::invalid(
            "invalid thinking level: must be one of off, low, medium, high, xhigh, max",
        )),
    }
}

/// The modification time, or the zero time when the platform has none.
pub(crate) fn modified_time(metadata: &Metadata) -> DateTime<Utc> {
    metadata
        .modified()
        .ok()
        .map(DateTime::<Utc>::from)
        .unwrap_or_else(kite_core::model::zero_time)
}

/// Identity, mode, size and mtime all match.
pub(crate) fn same_list_candidate_metadata(expected: &Metadata, current: &Metadata) -> bool {
    use std::os::unix::fs::MetadataExt;
    fsops::file_identity(expected) == fsops::file_identity(current)
        && expected.mode() == current.mode()
        && expected.len() == current.len()
        && modified_time(expected) == modified_time(current)
}

fn increment_skipped(count: i64) -> i64 {
    count.saturating_add(1)
}
